controllers/cassandrarestore_controller.go

```diff
@@ -52,221 +53,171 @@ type CassandraRestoreReconciler struct {
 // +kubebuilder:rbac:groups=cassandra.k8ssandra.io,namespace="medusa-operator",resources=cassandrarestores,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=cassandra.k8ssandra.io,namespace="medusa-operator",resources=cassandrarestores/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=cassandra.k8ssandra.io,namespace="medusa-operator",resources=cassandrarebackups,verbs=get;list;watch
-// +kubebuilder:rbac:groups=cassandra.datastax.com,namespace="medusa-operator",resources=cassandradatacenters,verbs=get;list;watch;create;update
+// +kubebuilder:rbac:groups=cassandra.datastax.com,namespace="medusa-operator",resources=cassandradatacenters,verbs=get;list;watch;create;update;patch
+// +kubebuilder:rbac:groups=apps,namespace="medusa-operator",resources=statefulsets,verbs=list;watch
 
 func (r *CassandraRestoreReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	ctx := context.Background()
-	_ = r.Log.WithValues("cassandrarestore", req.NamespacedName)
 
-	instance := &api.CassandraRestore{}
-	err := r.Get(ctx, req.NamespacedName, instance)
-	if err != nil {
-		if errors.IsNotFound(err) {
-			return ctrl.Result{}, nil
-		}
-		return ctrl.Result{RequeueAfter: 10 * time.Second}, err
+	factory := reconcile.NewFactory(r.Client, r.Log)
+	request, result, err := factory.NewRestoreRequest(ctx, req.NamespacedName)
+
+	if result != nil {
+		return *result, err
 	}
 
-	restore := instance.DeepCopy()
+	request.SetRestoreStartTime(metav1.Now())
+	request.SetRestoreKey(uuid.New().String())
 
-	if len(restore.Status.RestoreKey) == 0 {
-		if err = r.setRestoreKey(ctx, restore); err != nil {
-			// Could be stale item, we'll just requeue - this process can be repeated
-			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	if request.Restore.Spec.Shutdown && request.Restore.Status.DatacenterStopped.IsZero() {
+		if stopped := stopDatacenter(request); !stopped {
+			return r.applyUpdatesAndRequeue(ctx, request)
```

Comment:
And change the restoreKey. They're not protected in any case, but each reconcile loop changes them?

Reply:
`applyUpdatesAndRequeue()` checks to see if the CassandraRestore status has changed. If it has then the status is patched. It then checks to see if the CassandraDatacenter spec has changed. If it has, then the CassandraDatacenter is patched. If there are no changes, no updates should be made.

Reply:
Actually only in the case that SetRestoreStartTime did not actually check that if it was not changed. I don't really like this that the process isn't visible at all, but all rely on hidden side-effects which then create more side-effects. If applyUpdates doesn't applyUpdates because SetRestoreStartTime did not do SetRestoreStartTime, then that's kinda messy approach.

Reply:
>Since due to the previous "request.SetRestoreStartTime(..)" we keep updating that one with a new time all the time.
>
>And change the restoreKey. They're not protected in any case, but each reconcile loop changes them?

`SetRestoreStartTime()` and `SetRestoreStartTime()` are both idempotent. I added docs for both functions that mention this. Once they are persisted in the status, they won't get changed in subsequent reconcile loops.

>I don't really like this that the process isn't visible at all, but all rely on hidden side-effects which then create more side-effects. If applyUpdates doesn't applyUpdates because SetRestoreStartTime did not do SetRestoreStartTime, then that's kinda messy approach.

Would you find it more clear if I renamed `SetRestoreStartTime` to something like `CheckRestoreStartTime` or `SetRestoreStartTimeIfNecessary`?

I certainly could implement the logic directly in the `Reconcile` function along the following lines:

```go
if restore.Status.StartTime.IsZero() {
    restore.Status.StartTime = metav1.Now()
}

// continue with other reconciliation stuff....

if restoreStatusUpdated() {
    updateRestoreStatus()
}
```

That's where we are with the code in master and it is a mess. The additional abstraction of `SetRestoreStartTime` makes it easier to unit test that logic.

As for `applyUpdates` it simply checks to see if whether the status has changed. If it has, then it patches the status. Compared to master, I think this is a lot simpler and clearer. There is far less branching logic. Writes to the api server are limited to `applyUpdates` vs being scattered in multiple places. Lastly, the changes facilitate a better separation of concerns. The reconcile logic can focus on what changes it needs to make to the CassandraRestore and CassandraDatacenter. The `applyUpdates` function focuses on persisting those state changes.